File: leetcode/cpp/PassedProblems/237-delete-node-in-a-linked-list.cpp
/*
 * 237-delete-node-in-a-linked-list.cpp
 * Write a function to delete a node (except the tail) in a singly linked list,
 *  given only access to that node.
 *  Created on: Oct 3, 2015
 *      Author: hanzhe
 */


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <iostream>
using namespace std;
struct ListNode {
      int val;
      ListNode *next;
      ListNode(int x) : val(x), next(NULL) {}
  };
class Solution {
public:
    void deleteNode(ListNode* node) {
    	if(node == NULL)
    		return;
    	node->val = node->next->val;
    	node->next = node->next->next;
    }
};

File: leetcode/python/237-delete-node-in-a-linked-list.py
#!/usr/bin/env python
# encoding: utf-8


"""

Write a function to delete a node (except the tail) in a singly linked list, given only access to that node.

Supposed the linked list is 1 -> 2 -> 3 -> 4 and you are given the third node with value 3, the linked list should become 1 -> 2 -> 4 after calling your function.

"""


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 1次过
class Solution(object):
    def deleteNode(self, node):
        """
        :type node: ListNode
        :rtype: void Do not return anything, modify node in-place instead.
        """
        nn = node.next.next
        node.val = node.next.val
        node.next = nn

